### PR TITLE
Best Match Search Sort

### PR DESCRIPTION
**Pull request by eanders**
## Description

This brings the main client search more in-line with the HMIS search and implements the Most Relevant search ordering.

## Type of change
- [ ] Bug fix
- [x] New feature (adds functionality)
- [ ] Code clean-up / housekeeping
- [ ] Dependency update

## Checklist before requesting review
- [x] I have performed a self-review of my code
- [x] I have run the code that is being changed under ideal conditions, and it doesn't fail
- [x] My code includes comments and/or descriptive variable names to help other engineers understand the intent (or not applicable)
- [x] My code follows the style guidelines of this project (rubocop)
- [ ] I have updated the documentation (or not applicable)
- [ ] If it's not obvious how to test this change, I have provided testing instructions in this PR or the related issue

### CODE REVIEW

File: drivers/client_access_control/app/views/client_access_control/clients/search/_sort.haml [hunk 1]
```diff
@@ -4,9 +4,9 @@
...
-        - if @column == opt[:column] && @direction == opt[:direction]
-          - title = "<span class='icon-checkbox-checked'></span> #{opt[:title]}".html_safe
+      - GrdaWarehouse::Hud::Client::SORT_OPTIONS.each do |key, title|
+        - html = if @selected_sort == key
+          - "<span class='icon-checkbox-checked'></span> #{title}".html_safe
         - else
-          - title = "<span class='icon-checkbox-unchecked'></span> #{opt[:title]}".html_safe
-        %li= sort_as_link(title, opt[:column], opt[:direction])
+          - "<span class='icon-checkbox-unchecked'></span> #{title}".html_safe
+        %li= link_to(html, { q: params[:q], sort: key })
```
Comment (ttoomey) [inline drivers/client_access_control/app/views/client_access_control/clients/search/_sort.haml]: This is fine but maybe it's a good opportunity to tidy it up
```haml
      - GrdaWarehouse::Hud::Client::SORT_OPTIONS.each do |key, title|
        %li
          = link_to({ q: params[:q], sort: key }) do
            %span{class: @selected_sort == key ? 'icon-checkbox-checked' : 'icon-checkbox-unchecked'}
            = title
```

File: app/models/grda_warehouse/hud/client.rb [hunk 2]
```diff
@@ -1573,19 +1573,38 @@ def staff_types
...
+      when :first_name_a_to_z
+        order(arel_table[:FirstName].asc.nulls_last)
+      when :first_name_z_to_a
+        order(arel_table[:FirstName].desc.nulls_last)
+      when :age_youngest_to_oldest
+        order(arel_table[:DOB].desc.nulls_last)
+      when :age_oldest_to_youngest
+        order(arel_table[:DOB].asc.nulls_last)
+      else
+        raise NotImplementedError
```
Comment (ttoomey) [inline app/models/grda_warehouse/hud/client.rb]: maybe this is too fine a point but I'm not sure [NotImplementedError](https://ruby-doc.org/core-3.1.0/NotImplementedError.html) fits this use case. Maybe ArgumentError would be more appropriate (or just a simple raise)

```suggestion
        raise ArgumentError.new("invalid sort option #{option.inspect}")
```

File: app/models/grda_warehouse/hud/client.rb [hunk 3]
```diff
@@ -1573,19 +1573,38 @@ def staff_types
...
+      last_name_z_to_a: 'Last name Z-A',
+      first_name_a_to_z: 'First name A-Z',
+      first_name_z_to_a: 'First name Z-A',
+      age_youngest_to_oldest: 'Youngest to Oldest',
+      age_oldest_to_youngest: 'Oldest to Youngest',
+    }.freeze
+
+    def self.sort_by_option(option)
+      option = option&.to_sym
+      raise NotImplementedError unless SORT_OPTIONS.key?(option)
```
Comment (ttoomey) [inline app/models/grda_warehouse/hud/client.rb]: is this needed? The case statement seems like it's handling validation already

File: drivers/client_access_control/app/controllers/client_access_control/search_concern.rb [hunk 4]
```diff
@@ -73,12 +61,11 @@ def sort_filter_index
...
+      sort_option = params[:sort].presence || :best_match
+      @clients = @clients.sort_by_option(sort_option)
 
-      @clients = @clients.order(*sort) if sort.any?
-
-      @column = sort_column
-      @direction = sort_direction
-      @sort_columns = client_sort_columns + client_processed_sort_columns
+      @selected_sort = params[:sort]&.to_sym.presence || :best_match
+      @sort_title = GrdaWarehouse::Hud::Client::SORT_OPTIONS[@selected_sort]
```
Comment (ttoomey) [inline drivers/client_access_control/app/controllers/client_access_control/search_concern.rb]: small suggestions:
* not sure we need both `sort_option` and `@selected_sort`
```ruby
      @selected_sort = params[:sort].presence&.to_sym || :best_match
      @clients = @clients.sort_by_option(@selected_sort)
      @sort_title = GrdaWarehouse::Hud::Client::SORT_OPTIONS[@selected_sort]
```

**Review by ttoomey — APPROVED** [review]
Looks great to me, thanks!

File: app/models/grda_warehouse/hud/client.rb (hunk 3 above)
Comment (eanders) [reply app/models/grda_warehouse/hud/client.rb]: I'll remove, just copied these from the HMIS version

File: drivers/client_access_control/app/controllers/client_access_control/search_concern.rb (hunk 4 above)
Comment (eanders) [reply drivers/client_access_control/app/controllers/client_access_control/search_concern.rb]: Thanks, missed that.

File: drivers/client_access_control/app/views/client_access_control/clients/search/_sort.haml (hunk 1 above)
Comment (eanders) [reply drivers/client_access_control/app/views/client_access_control/clients/search/_sort.haml]: Thanks!

**Pull request merged**